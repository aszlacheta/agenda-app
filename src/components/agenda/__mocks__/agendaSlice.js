const agendaSlice = {
  removeEntry: jest.fn(),
  setNonEditMode: jest.fn()
};

module.exports = agendaSlice;
