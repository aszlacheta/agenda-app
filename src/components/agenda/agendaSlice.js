import { createSlice } from '@reduxjs/toolkit';

const compareEntry = (a, b) => {
  if (a.startDate < b.startDate) {
    return -1;
  }
  if (a.startDate > b.startDate) {
    return 1;
  }
  return 0;
};

export const agendaSlice = createSlice({
  name: 'agenda',
  initialState: {
    entries: [],
    id: '',
    day: 0
  },
  reducers: {
    setAgenda: (state, { payload }) => {
      const { id, day, entries } = payload;
      state.id = id;
      state.day = day;
      state.entries = entries;
    },
    setEntries: (state, { payload }) => {
      state.entries = payload;
    },
    addEntry: (state, { payload }) => {
      payload.isEdited = true;
      state.entries.push(payload);
    },
    removeEntry: (state, { payload }) => {
      state.entries = state.entries.filter(entry => entry.id !== payload);
    },
    setEditMode: (state, { payload: entryId }) => {
      state.entries = state.entries.map(entry => {
        if (entry.id === entryId) {
          entry.isEdited = true;
        }
        return entry;
      });
    },
    setNonEditMode: (state, { payload: entryId }) => {
      state.entries = state.entries.map(entry => {
        if (entry.id === entryId) {
          entry.isEdited = false;
        }
        return entry;
      });
    },
    saveEntry: (state, { payload }) => {
      const { entryId, entry } = payload;
      const index = state.entries.findIndex(entry => entry.id === entryId);

      Object.keys(entry).forEach((key) => {
        state.entries[index][key] = entry[key];
      });

      state.entries[index].editable = true;
      state.entries[index].isNew = undefined;
      state.entries[index].isEdited = false;

      state.entries.sort(compareEntry);
    }
  }
});

export const { setAgenda, addEntry, removeEntry, setEntries, saveEntry, setEditMode, setNonEditMode } = agendaSlice.actions;

export default agendaSlice.reducer;
