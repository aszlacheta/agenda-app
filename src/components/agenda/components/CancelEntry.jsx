import React from 'react';
import { useDispatch } from 'react-redux';
import { removeEntry, setEdited } from "../agendaSlice";

export default function CancelEntry({ entryId, isNew }) {
    const dispatch = useDispatch();

    const onClick = () => {
        if (isNew) {
            dispatch(removeEntry(entryId));
        } else {
            dispatch(setEdited({ entryId, isEdited: false }))
        }
    }

    return <button className="cancel-entry" onClick={onClick}>cancel</button>
}