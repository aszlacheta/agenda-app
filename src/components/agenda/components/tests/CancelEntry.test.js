import 'jsdom-global/register';
import React from 'react';
import CancelEntry from '../CancelEntry';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { removeEntry, setNonEditMode } from '../../agendaSlice';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../../agendaSlice');

describe('Cancel entry component', () => {
  const entryId = 'testEntryId';

  it('should render properly', () => {
    const tree = renderer
      .create(<CancelEntry entryId={entryId} isNew={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should remove entry if it was newly created', () => {
    const component = mount(<CancelEntry entryId={entryId} isNew={true} />);
    const button = component.find('button');

    button.simulate('click');

    expect(removeEntry).toHaveBeenCalled();
  });

  it('should cancel editing if it was already saved', () => {
    const component = mount(<CancelEntry entryId={entryId} isNew={false} />);
    const button = component.find('button');

    button.simulate('click');

    expect(setNonEditMode).toHaveBeenCalled();
  });
});
