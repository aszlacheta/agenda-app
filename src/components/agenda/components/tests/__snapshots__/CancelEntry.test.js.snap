// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cancel entry component should render properly 1`] = `
<button
  className="cancel-entry"
  data-cy="cancel-button"
  onClick={[Function]}
>
  entry.buttons.cancel, 
</button>
`;
